### PR TITLE
Coveralls based coverage testing

### PR DESCRIPTION
**Pull request by konserw**
Code coverage solution based on gcov and Coveralls service: https://coveralls.io/github/konserw/cucumber-cpp
Of course badge and api key is to be replaced.
Fixes issue #123

### CODE REVIEW

**Comment by muggenhor** [comment]
This is a nice idea, but instead of hacking this into our regular build system I would prefer an approach that changes only the CI system. With only very minimal changes to the build system (at the most I would expect the build system to produce a list of relevant source files).

What I would expect instead:

1. modifying the CI script to add the desired [instrumentation options](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html) to CMake's invocation (probably: `-DCMAKE_CXX_FLAGS='--coverage' -DCMAKE_LINK_INTERFACE_LIBRARIES='gcov'`).
2. adding a post-processing step that collects the desired output after test execution and transmits it to coveralls

The first bit is probably just adding this to travis.sh's CMake invocation:
```sh
${COVERALLS_SERVICE_NAME:+${COVERALLS_REPO_TOKEN:+-DCMAKE_CXX_FLAGS='--coverage' -DCMAKE_LINK_INTERFACE_LIBRARIES='gcov'}}
```

The second bit is more involved but should be similar to this pseudo code:
```sh
if [ -n "${COVERALLS_SERVICE_NAME:-}" -a "${COVERALLS_REPO_TOKEN:-}" ]; then
    # Let gcov extract coverage data
    cmake -E make_directory build/gcov
    (
        cd build/gcov
        for f in `find .. -name '*.gcda' -type f`; do
            gcov -p -o "$(dirname "$f")" "$f"
        done
    ) > /dev/null

    # Translate _relevant_ gcov files to a format coveralls can consume
    for f in build/gcov/*\#cucumber-cpp\#src\#* build/gcov/*\#include\#cucumber-cpp\#*; do
        convert_from_gcov_to_coveralls "$f"
    done

    transmit_coverage_data_to_coveralls
fi
```

I expect none of that needs changes to the build system. The only exception would be if we would want to have some of the build system's info available to us to determine which files are relevant to display coverage for (instead of the very simple, but effective, glob patterns I listed).

**Comment by konserw** [comment]
@muggenhor @paoloambrosio @aslakhellesoy - could you create cucumber account on coveralls ?

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -33,6 +33,9 @@ if(CMAKE_EXTRA_GENERATOR OR MSVC_IDE)
     list(APPEND CUKE_SOURCES ${CUKE_HEADERS})
 endif()
 
+string(REGEX REPLACE "([^;]+)" "${CMAKE_CURRENT_SOURCE_DIR}/\\1" COVER_SOURCES "${CUKE_SOURCES}")
+file(WRITE sourceFileList "${COVER_SOURCES}" )
```
Comment (konserw) [inline src/CMakeLists.txt]: It was just a thing I was playing with and even deleted from my repo - just forgot pushing deletion commit here

File: CMakeLists.txt [hunk 2]
```diff
@@ -14,7 +14,10 @@ set(GMOCK_SRC_DIR "" CACHE STRING "Google Mock framework sources path (otherwise
 set(GMOCK_VER "1.7.0" CACHE STRING "Google Mock framework version to be used")
 option(VALGRIND_TESTS "Run tests within Valgrind" OFF)
 
-set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake/modules)
+set(CMAKE_MODULE_PATH
+    ${CMAKE_MODULE_PATH} 
+    ${PROJECT_SOURCE_DIR}/cmake/modules
+)
```
Comment (konserw) [inline CMakeLists.txt]: Ok

File: travis.sh [hunk 3]
```diff
@@ -7,27 +7,35 @@ bundle install
 CTEST_OUTPUT_ON_FAILURE=ON
 export CTEST_OUTPUT_ON_FAILURE
 
+export COVERALLS_SERVICE_NAME=travis-ci
+export COVERALLS_REPO_TOKEN=ThdhpdagGJJVJj0KFGdSgdpzRLALbNAIg
```
Comment (konserw) [inline travis.sh]: We can try without token

File: CMakeLists.txt [hunk 4]
```diff
@@ -14,7 +14,7 @@ set(GMOCK_SRC_DIR "" CACHE STRING "Google Mock framework sources path (otherwise
 set(GMOCK_VER "1.7.0" CACHE STRING "Google Mock framework version to be used")
 option(VALGRIND_TESTS "Run tests within Valgrind" OFF)
 
-set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake/modules)
+list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake/modules)
```
Comment (muggenhor) [inline CMakeLists.txt]: I think this is causing your build failures. I guess `list(APPEND)` doesn't work for variables not yet defined.

**Comment by muggenhor** [comment]
So it looks like POSTing to coveralls works, but I don't think it manages to extract the coverage data:
>  **0** of **0** relevant lines covered (**NaN%**)

**Comment by konserw** [comment]
I don't understand this - build from the same branch/commit is fine on my account:
https://coveralls.io/github/konserw/cucumber-cpp

**Comment by muggenhor** [comment]
Well at least it's not the absence of the API token that's at fault, when that's included this is the result: https://coveralls.io/builds/12854344

**Comment by konserw** [comment]
> I don't understand this - build from the same branch/commit is fine on my account:
https://coveralls.io/github/konserw/cucumber-cpp

Please disregard that comment - I was looking at wrong branch. With this branch I've also got 0 relevant lines.

**Comment by konserw** [comment]
I don't know how to fix this implementation. I'll create new pull request with different approach.

**Comment by muggenhor** [comment]
FYI: this is continued in #162

**Pull request closed without merging**